examples/nlp/mlm_with_bert.py

```diff
@@ -0,0 +1,272 @@
+        # set labels for masked tokens
+        labels[inp_mask] = X[inp_mask]
+
+        # prepare input
+        X_mlm = np.copy(X)
+        # set input to [MASK] which is the last token for the 90% of tokens
+        # this means leaving 10% unchanged
+        inp_mask_2mask = inp_mask & (np.random.rand(*X.shape) < 0.90)
+        X_mlm[
+            inp_mask_2mask
+        ] = self.tokenizer.mask_token_id  # mask token is the last in the dict
+
+        # set 10% to a random token
+        inp_mask_2random = inp_mask_2mask & (np.random.rand(*X.shape) < 1 / 9)
+        X_mlm[inp_mask_2random] = np.random.randint(
+            3, self.tokenizer.mask_token_id, inp_mask_2random.sum()
+        )
+
+        return X_mlm, labels
+
+    def create_dist_dataset(self, X, y=None, training=False):
+
+        dataset = tf.data.Dataset.from_tensor_slices(X)
+
+        ### Add y if present ###
+        if y is not None:
+            dataset_y = tf.data.Dataset.from_tensor_slices(y)
+            dataset = tf.data.Dataset.zip((dataset, dataset_y))
+
+        ### Repeat if training ###
+        if training:
+            dataset = dataset.shuffle(len(X)).repeat()
+
+        dataset = dataset.batch(global_batch_size).prefetch(AUTO)
+
+        ### make it distributed  ###
+        dist_dataset = self.strategy.experimental_distribute_dataset(dataset)
+
+        return dist_dataset
+
+
+"""shell
+wget https://raw.githubusercontent.com/SrinidhiRaghavan/AI-Sentiment-Analysis-on-IMDB-Dataset/master/imdb_tr.csv
+"""
+
+data = pd.read_csv("imdb_tr.csv", encoding="ISO-8859-1")
+tokenizer = AutoTokenizer.from_pretrained(flags.PRETRAINED_MODEL)
+dataset = Dataset(tokenizer, strategy)
+texts = data.text.values
+train_dist_dataset = dataset.get_processed_dataset(texts, flags.MAX_LEN)
+
+
+"""
+## Create MaskedLanguageModel using huggingface transformers
+"""
+
+
+class MaskedLanguageModel:
+    def __init__(self, strategy, model_name):
+
+        self.strategy = strategy
+        with self.strategy.scope():
+            self.model = TFAutoModelWithLMHead.from_pretrained(model_name)
+            self.optimizer = tf.keras.optimizers.Adam(learning_rate=flags.LR)
+
+        (
+            self.compute_mlm_loss,
+            self.train_mlm_loss_metric,
+        ) = self.get_mlm_loss_and_metrics()
+
+    def get_mlm_loss_and_metrics(self):
+
+        with self.strategy.scope():
+            mlm_loss_object = self.masked_sparse_categorical_crossentropy
+
+            def compute_mlm_loss(labels, predictions):
+                per_example_loss = mlm_loss_object(labels, predictions)
+                loss = tf.nn.compute_average_loss(
+                    per_example_loss, global_batch_size=global_batch_size
+                )
+                return loss
+
+            train_mlm_loss_metric = tf.keras.metrics.Mean()
+
+        return compute_mlm_loss, train_mlm_loss_metric
+
+    def masked_sparse_categorical_crossentropy(self, y_true, y_pred):
```

Comment:
I have tried using SparseCategoricalCrossentropy sample_weight argument but it's giving nan loss (because I have used -1 (negative/ unmasked token) in labels tensor). Should I use non-negative id for unmasked token and then use sample_weight?